src/backend/distributed/transaction/distributed_deadlock_detection.c

```diff
@@ -0,0 +1,303 @@
+	Datum values[attributeCount];
+	bool isNulls[attributeCount];
+
+	CheckCitusVersion(ERROR);
+
+	/* check to see if caller supports us returning a tuplestore */
+	if (returnSetInfo == NULL || !IsA(returnSetInfo, ReturnSetInfo))
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("set-valued function called in context " \
+						"that cannot accept a set")));
+	}
+
+	if (!(returnSetInfo->allowedModes & SFRM_Materialize))
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("materialize mode required, but it is not " \
+						"allowed in this context")));
+	}
+
+	/* build a tuple descriptor for our result type */
+	if (get_call_result_type(fcinfo, NULL, &tupleDescriptor) != TYPEFUNC_COMPOSITE)
+	{
+		elog(ERROR, "return type must be a row type");
+	}
+
+	perQueryContext = returnSetInfo->econtext->ecxt_per_query_memory;
+
+	oldContext = MemoryContextSwitchTo(perQueryContext);
+
+	tupleStore = tuplestore_begin_heap(true, false, work_mem);
+	returnSetInfo->returnMode = SFRM_Materialize;
+	returnSetInfo->setResult = tupleStore;
+	returnSetInfo->setDesc = tupleDescriptor;
+
+	MemoryContextSwitchTo(oldContext);
+
+	waitGraph = BuildGlobalWaitGraph();
+	distributedTransactionHash = FlattenWaitGraph(waitGraph);
+
+	/* iterate on all nodes */
+	hash_seq_init(&status, distributedTransactionHash);
+
+	while ((transctionNode = (TransactionNode *) hash_seq_search(&status)) != 0)
+	{
+		memset(values, 0, sizeof(values));
+		memset(isNulls, false, sizeof(isNulls));
+
+		values[0] = UInt64GetDatum(transctionNode->transactionId.transactionNumber);
+		values[1] = CStringGetDatum(WaitsForToString(transctionNode->waitsFor));
+
+		tuplestore_putvalues(tupleStore, tupleDescriptor, values, isNulls);
+	}
+
+	/* clean up and return the tuplestore */
+	tuplestore_donestoring(tupleStore);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * WaitsForToString is only intended for testing and debugging. It gets a
+ * waitsForList and returns the list of transaction nodes' transactionNumber
+ * in a string.
+ */
+static char *
+WaitsForToString(List *waitsFor)
+{
+	StringInfo transactionIdStr = makeStringInfo();
+	ListCell *waitsForCell = NULL;
+
+	foreach(waitsForCell, waitsFor)
+	{
+		TransactionNode *waitingNode = (TransactionNode *) lfirst(waitsForCell);
+
+		if (transactionIdStr->len != 0)
+		{
+			appendStringInfoString(transactionIdStr, ",");
+		}
+
+		appendStringInfo(transactionIdStr, "%ld",
+						 waitingNode->transactionId.transactionNumber);
+	}
+
+	return transactionIdStr->data;
+}
+
+
+/*
+ * FlattenWaitGraph converts a wait graph which is represented as a set of
+ * wait edges between all backends in the Citus cluster, to a set of nodes,
+ * one for each distributed transaction.
+ *
+ * The returned structure is an adjacency list with the following format:
+ * [transactionId] = [transactionNode->waitsFor {list of waiting transaction nodes}]
+ */
+static HTAB *
+FlattenWaitGraph(WaitGraph *waitGraph)
```

Comment:
We could name this something more descriptive, thinking out loud: `BuildDistributedTransactionWaitsForHash`. 

Also wondering: Should this be in `lock_graph.c`? In that case we could even make WaitGraph private.